crates/schema/src/def.rs

```diff
@@ -0,0 +1,640 @@
+    pub fn stored_in_table_def(&self, name: &Identifier) -> Option<&TableDef> {
+        self.stored_in_table_def
+            .get(name)
+            .and_then(|table_name| self.tables.get(table_name))
+    }
+
+    /// Lookup a definition by its key in `self`.
+    pub fn lookup<T: ModuleDefLookup>(&self, key: T::Key<'_>) -> Option<&T> {
+        T::lookup(self, key)
+    }
+
+    /// Generate indexes for the module definition.
+    /// We guarantee that all `unique` constraints have an index generated for them.
+    /// This will be removed once another enforcement mechanism is implemented.
+    /// We implement this after validation to share logic between v8 and v9.
+    fn generate_indexes(&mut self) {
+        for table in self.tables.values_mut() {
+            for constraint in table.unique_constraints.values() {
+                // if we have a constraint for the index, we're fine.
+                if table.indexes.values().any(|index| {
+                    let IndexDef {
+                        algorithm: IndexAlgorithm::BTree { columns },
+                        ..
+                    } = index;
+
+                    columns == &constraint.columns
+                }) {
+                    continue;
+                }
+
+                let column_names = constraint
+                    .columns
+                    .iter()
+                    .map(|col_id| &*table.get_column(col_id).expect("validated unique constraint").name)
+                    .join("_");
+
+                // TODO(1.0): ensure generated index names are identical when upgrading the Rust module bindings.
+                let mut index_name =
+                    Identifier::new(format!("idx_{}_{}_{}_unique", table.name, column_names, constraint.name).into())
+                        .expect("validated identifier parts");
+
+                // incredibly janky loop to avoid name collisions.
+                // hey, somebody could be being malicious.
+                while self.stored_in_table_def.contains_key(&index_name) {
+                    index_name =
+                        Identifier::new(format!("{}_1", index_name).into()).expect("validated identifier parts");
+                }
+
+                table.indexes.insert(
+                    index_name.clone(),
+                    IndexDef {
+                        name: index_name.clone(),
+                        algorithm: IndexAlgorithm::BTree {
+                            columns: constraint.columns.clone(),
+                        },
+                        accessor_name: None, // this is a generated index.
+                    },
+                );
+                self.stored_in_table_def.insert(index_name, table.name.clone());
+            }
+        }
+    }
+}
+
+impl TryFrom<RawModuleDef> for ModuleDef {
+    type Error = ValidationErrors;
+
+    fn try_from(raw: RawModuleDef) -> Result<Self, Self::Error> {
+        Self::validate(raw)
```

Comment:
Nit but I'd just inline the `validate` implementation here; seems pointless to have custom `ModuleDef::validate` when a standard API will suffice.